rclcpp_action/src/server.cpp

```diff
@@ -334,11 +337,17 @@ ServerBase::execute_goal_request_received(std::shared_ptr<void> & data)
     // Copy out goal handle since action server storage disappears when it is fini'd
     *handle = *rcl_handle;
 
-    pimpl_->goal_handles_[uuid] = handle;
+    {
+      std::lock_guard<std::recursive_mutex> lock(pimpl_->action_server_reentrant_mutex_);
+      pimpl_->goal_handles_[uuid] = handle;
+    }
```

Comment:
I don't think we need this particular lock here.  At least, this lock is meant to protect the `action_server`, and that is not being accessed here.

Reply:
https://github.com/ros2/rclcpp/blob/6feedbe117cc163a5b7f7d4d66a3446445ea6b5e/rclcpp_action/src/server.cpp#L526

goal_handles_ can be changed simultaneously, so I put a lock guard here too.

Reply:
I'm sorry to be a bit pedantic here, but the name of the lock is currently `action_server_reentrant_mutex`.  As a reader of the code, then, I expect that that lock will protect the `action_server`, and not necessarily protect anything else.

But I do agree that the `goal_handles` could be changed by two separate threads.  There are 2 ways we can go with this:
1.  Rename the lock to `pimpl_lock` or something like that.  At that point, it makes more sense that the lock protects the entire PIMPL structure.  But we would also need to audit the rest of the code to make sure that the lock is being held anytime anything in the structure is being modified.
2. Introduce another lock specifically to protect the `goal_handles`, and lock that as appropriate.

I'm honestly not sure which is better.

Reply:
I missed this message and have made a change, but I added another mutex for unordered maps including goal_handles, because the pimpl structure has three groups of variables, 1) action_server and its attributes (size_t, bool) captured from actsion_server, 2) three unordered maps, and 3) references (clock and logger).
I think we don't need to lock the entire pimpl structure.

One line I'm not sure about is updating goal_request_ready_ without a lock.
All size_t and bool variables are protected because they are updated with action_server.
https://github.com/ros2/rclcpp/blob/5ae0e9a9929138a05adc40d0bfb448fb5816e819/rclcpp_action/src/server.cpp#L292

Reply:
> One line I'm not sure about is updating goal_request_ready_ without a lock.

Good question.  It *should* be atomic since it is just updating a boolean, but that is not guaranteed.  One thing we could do there is make it a `std::atomic<bool>`, which would guarantee it.

Reply:
Thanks for the suggestion! I changed it to `std::atomic<bool>`.